clients/src/main/java/org/apache/kafka/clients/admin/KafkaAdminClient.java

```diff
@@ -736,42 +794,27 @@ final void fail(long now, Throwable throwable) {
                 runnable.enqueue(this, now);
                 return;
             }
-            tries++;
-            nextAllowedTryMs = now + retryBackoffMs;
 
             // If the call has timed out, fail.
             if (calcTimeoutMsRemainingAsInt(now, deadlineMs) < 0) {
                 failWithTimeout(now, throwable);
                 return;
             }
-            // If the exception is not retriable, fail.
-            if (!(throwable instanceof RetriableException)) {
-                if (log.isDebugEnabled()) {
-                    log.debug("{} failed with non-retriable exception after {} attempt(s)", this, tries,
-                        new Exception(prettyPrintException(throwable)));
-                }
-                handleFailure(throwable);
-                return;
-            }
-            // If we are out of retries, fail.
-            if (tries > maxRetries) {
-                failWithTimeout(now, throwable);
-                return;
-            }
+
             if (log.isDebugEnabled()) {
-                log.debug("{} failed: {}. Beginning retry #{}",
-                    this, prettyPrintException(throwable), tries);
+                log.debug("{} failed with non-retriable exception after {} attempt(s)", this,
+                        getRetryContext().getTries(), new Exception(prettyPrintException(throwable)));
             }
-            runnable.enqueue(this, now);
+            handleFailure(throwable);
         }
 
         private void failWithTimeout(long now, Throwable cause) {
             if (log.isDebugEnabled()) {
-                log.debug("{} timed out at {} after {} attempt(s)", this, now, tries,
+                log.debug("{} timed out at {} after {} attempt(s)", this, now, retryContext.getTries() + 1,
```

Comment:
Wouldn't this just be `retryContext.tries()` and not `retryContext.tries() + 1`?

Reply:
I think attempts (total tries) = retries + 1. The retryContext.tries() records the time of retries, not the total number of attempts.

Reply:
Ah okay makes sense.